### PR TITLE
fix ansible remediations to avoid creating duplicate entries

### PR DESCRIPTION
**Pull request by vojtapolasek**
#### Description:

Ansible remediations for audit_rules_kernel_module_loading and audit_rules_media_export are modified to check if needed Audit lines already exist. Therefore, they should not create duplicates.
There is also new test checking for that in audit_rules_kernel_module_loading, as the remediation is quite complex.

#### Rationale:

avoid duplicate Audit rules

### CODE REVIEW

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/tests/syscalls_one_per_line_one_missing.fail.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+# profiles = xccdf_org.ssgproject.content_profile_C2S
+# remediation = bash
```
Comment (yuumasato) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/tests/syscalls_one_per_line_one_missing.fail.sh:3]: Remediation metadata should be removed, this is disabling testing with Ansible in Test Suite.
Other test scenarios should have this line removed as well.

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml [hunk 2]
```diff
@@ -34,48 +123,197 @@
...
+- name: add delete_module into line for 32 bit rules.d
+  set_fact:
+    audit_kernel_line_32_rules_d= {{ audit_kernel_line_32_rules_d + '-S delete_module ' }}
+  when: find_existing_kernel_delete_module_32_rules_d is defined and find_existing_kernel_delete_module_32_rules_d.matched == 0 and audit_kernel_line_32_rules_d is defined
+
+{{% if product != "rhel6" %}}
+- name: add finit_module into line for 32 bit rules.d
+  set_fact:
+    audit_kernel_line_32_rules_d= {{ audit_kernel_line_32_rules_d + '-S finit_module ' }}
+  when: find_existing_kernel_finit_module_32_rules_d is defined and find_existing_finit_delete_module_32_rules_d.matched == 0 and audit_kernel_line_32_rules_d is defined
```
Comment (yuumasato) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml:153]: There is a typo, and it says that `find_existing_finit_delete_module_32_rules_d` is undefined.

**Comment by vojtapolasek** [comment]
Thank you for the review, yes I know that the solution is not ideal. I am working on rewriting this to loops.

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml [hunk 3]
```diff
@@ -35,47 +94,78 @@
...
+  block:
+    - name: start the line
+      set_fact: tmpline="-a always,exit -F arch=b64 "
+    - name: add syscalls
+      set_fact: tmpline="{{tmpline + '-S ' + item.item + ' ' }}"
+      loop: "{{ audit_kernel_found_64_rules_d.results }}"
+      when: item.matched is defined and item.matched == 0
+    - name: finish the line
+      set_fact: tmpline="{{ tmpline + '-k modules' }}"
+    - name: insert/replace the line in appropriate file
```
Comment (yuumasato) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml:124]: Watch out for the capitalization of the task name.
The task only adds new audit rules, it doesn't replace them.

The name of task here can be a bit more descriptive like in the block name.

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml [hunk 4]
```diff
@@ -35,47 +94,78 @@
...
-    path: "{{ all_files[0] }}"
-    {{% if product == "rhel6" %}}
-    line: "-a always,exit -F arch=b64 -S init_module -S delete_module -k modules"
-    {{% else %}}
-    line: "-a always,exit -F arch=b64 -S init_module -S delete_module -S finit_module -k modules"
-    {{% endif %}}
-    create: yes
-  when: audit_arch is defined and audit_arch == 'b64'
+  block:
+    - name: start the line
```
Comment (yuumasato) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml:116]: Construct rule: add rule list, action and arch

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml [hunk 5]
```diff
@@ -35,47 +94,78 @@
...
+      set_fact: tmpline="-a always,exit -F arch=b64 "
+    - name: add syscalls
+      set_fact: tmpline="{{tmpline + '-S ' + item.item + ' ' }}"
+      loop: "{{ audit_kernel_found_64_rules_d.results }}"
+      when: item.matched is defined and item.matched == 0
+    - name: finish the line
```
Comment (yuumasato) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml:122]: Construct rule: add key

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml [hunk 6]
```diff
@@ -35,47 +94,78 @@
...
-    line: "-a always,exit -F arch=b64 -S init_module -S delete_module -k modules"
-    {{% else %}}
-    line: "-a always,exit -F arch=b64 -S init_module -S delete_module -S finit_module -k modules"
-    {{% endif %}}
-    create: yes
-  when: audit_arch is defined and audit_arch == 'b64'
+  block:
+    - name: start the line
+      set_fact: tmpline="-a always,exit -F arch=b64 "
+    - name: add syscalls
```
Comment (yuumasato) [inline linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml:118]: Construct rule: add syscalls

File: linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml [hunk 7]
```diff
@@ -11,14 +11,73 @@
...
+      {{% endif %}}
+
+- name: declare number of syscalls
+  set_fact: audit_kernel_number_of_syscalls="{{ syscalls|length|int }}"
+
+
+- name: Check existence of syscalls for 32 bit architecture in /etc/audit/rules.d/*
+  find:
+    paths: "/etc/audit/rules.d"
+    contains: '^\s*-a\s+always,exit\s+-F\s+arch=b32\s+.*-S\s+.*{{ item }}.*$'
```
[the remark this answers is not in the document]
Comment (yuumasato) [reply linux_os/guide/system/auditing/auditd_configure_rules/audit_kernel_module_loading/audit_rules_kernel_module_loading/ansible/shared.yml:30]: This test scenario triggers it:
```
-a always,exit -F arch=b32 -S delete_module -S finit_module -k modules
-a always,exit -F arch=b64 -S delete_module -S finit_module -k modules
```

**Comment by yuumasato** [comment]
Great work!

**Pull request merged**